### PR TITLE
LLAMA3 SQLite Baseline with ASCII Tokenizer (adapting to HuggingFace)

### PR DESCRIPTION
**Pull request by savitha-eng**
### Description

This PR adds Phase 1 SQLite baseline in 3 logical components:

## Part 1: HuggingFace Tokenizer (Commit e1bf6789)
- PreTrainedTokenizerFast with NeMo convention
- 9 tokenizer tests

## Part 2: SQLite Dataset (Commit b5891937)  
- EVO2-style windowing
- 13 dataset/dataloader tests

## Part 3: Training Integration (Commit 3a8a1257)
- Updated train.py and model.py
- Validated training (loss decreasing)

#### Usage

```python
TODO: Add code snippet
```

### Type of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor
- [ ] Documentation update
- [ ] Other (please describe):

### CI Pipeline Configuration

Configure CI behavior by applying the relevant labels. By default, only basic unit tests are run.

- [ciflow:skip](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:skip) - Skip all CI tests for this PR
- [ciflow:notebooks](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:notebooks) - Run Jupyter notebooks execution tests for bionemo2
- [ciflow:slow](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:slow) - Run slow single GPU integration tests marked as @pytest.mark.slow for bionemo2
- [ciflow:all](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all) - Run all tests (unit tests, slow tests, and notebooks) for bionemo2. This label can be used to enforce running tests for all bionemo2.
- [ciflow:all-recipes](https://github.com/NVIDIA/bionemo-framework/blob/main/docs/docs/main/contributing/contributing.md#ciflow:all-recipes) - Run tests for all recipes (under bionemo-recipes). This label can be used to enforce running tests for all recipes.

Unit tests marked as `@pytest.mark.multi_gpu` or `@pytest.mark.distributed` are not run in the PR pipeline.

For more details, see [CONTRIBUTING](CONTRIBUTING.md)

> [!NOTE]
> By default, only basic unit tests are run. Add appropriate labels to enable an additional test coverage.

#### Authorizing CI Runs

We use [copy-pr-bot](https://docs.gha-runners.nvidia.com/apps/copy-pr-bot/#automation) to manage authorization of CI
runs on NVIDIA's compute resources.

- If a pull request is opened by a trusted user and contains only trusted changes, the pull request's code will
  automatically be copied to a pull-request/ prefixed branch in the source repository (e.g. pull-request/123)
- If a pull request is opened by an untrusted user or contains untrusted changes, an NVIDIA org member must leave an
  `/ok to test` comment on the pull request to trigger CI. This will need to be done for each new commit.

### Pre-submit Checklist

- [ ] I have tested these changes locally
- [ ] I have updated the documentation accordingly
- [ ] I have added/updated tests as needed
- [ ] All existing tests pass successfully

### CODE REVIEW

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+    cursor.execute("""
+        CREATE TABLE sequences (
+            contig_id TEXT PRIMARY KEY,
+            nt_sequence TEXT NOT NULL,
+            length INTEGER NOT NULL
+        )
+    """)
+    
+    test_sequences = [
+        ("seq_A", "A" * 10000, 10000),
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: you probably only need like 10 samples right? Can you just divide all these by 100?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    
+    conn.commit()
+    conn.close()
+    
+    yield db_path
+    
+    Path(db_path).unlink()
+
+
+def test_dataset_index_0_returns_all_as(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: def test_dataset_loads_sample_sequence might be better name

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    nucleotides = sample["input_ids"][1:-1]  # Remove BOS/EOS
+    
+    # With seed=42, index 0 is all As
+    expected_sequence = [65] * len(nucleotides)  # All As
+    received_sequence = nucleotides.tolist()
+    
+    assert received_sequence == expected_sequence
+
+
+def test_dataset_index_1_returns_all_ts(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: This behavior is already covered in the test above right?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    nucleotides = sample["input_ids"][1:-1]
+    
+    # With seed=42, index 1 is all Ts (100 tokens)
+    expected_sequence = [84] * 100  # All Ts
+    received_sequence = nucleotides.tolist()
+    
+    assert received_sequence == expected_sequence
+
+
+def test_dataset_index_3_returns_all_gs(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: The behavior under test here is already covered under the test for As. You can delete this test and all others that test the same behavior.

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    nucleotides = sample["input_ids"][1:-1]
+    
+    # With seed=42, index 3 is all Gs (100 tokens)
+    expected_sequence = [71] * 100  # All Gs
+    received_sequence = nucleotides.tolist()
+    
+    assert received_sequence == expected_sequence
+
+
+def test_dataset_index_5_returns_all_cs(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: delete

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    nucleotides = sample["input_ids"][1:-1]
+    
+    # With seed=42, index 5 is all Cs (100 tokens)
+    expected_sequence = [67] * 100  # All Cs
+    received_sequence = nucleotides.tolist()
+    
+    assert received_sequence == expected_sequence
+
+
+def test_dataloader_returns_batch_with_required_keys(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: change `required` --> `expected`

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+    
+    train_iterator, epoch_len = create_genomic_dataloader(args, tokenizer)
+    batch = next(train_iterator)
+    
+    assert "input_ids" in batch
+    assert "labels" in batch
+    assert "attention_mask" in batch
+
+
+def test_batch_size_is_correct(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: I think you can cover more ground if you simply have a test like this
1. Make some sample data
2. Make a variable called `expected_batch` and set it equal to what you think the batch should look like after grabbing data from the dataloader. This will include things like "batch_size", and other batch variables like "sequence_length" etc. 
3. Send the sample data through the dataloader and receive the `returned_batch`
4. Compare `assertEqual(expected_batch, returned_batch)` ((or some equivalent way of comparing dicts))
Then you'll have covered a lot of these tests swith a single test. In the test itself, when you are comparing things you can even be more verbose and raise ValueError's based on what failed

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    })
+    
+    train_iterator, epoch_len = create_genomic_dataloader(args, tokenizer)
+    batch = next(train_iterator)
+    
+    padding_mask = batch["attention_mask"] == 0
+    assert torch.all(batch["labels"][padding_mask] == -100)
+
+
+def test_epoch_length_equals_total_windows_divided_by_batch_size(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py:216]: good test!

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+    
+    train_iterator, epoch_len = create_genomic_dataloader(args, tokenizer)
+    batch = next(train_iterator)
+    
+    # For real tokens, labels should equal input_ids
+    real_mask = batch["attention_mask"] == 1
+    assert torch.all(batch["labels"][real_mask] == batch["input_ids"][real_mask])
+
+
+def test_padding_positions_are_minus_100(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: This test is somewhat unclear. 

1. It looks like you're checking if your label id is -100 for non masked values. But since you're not explicilty setting that value in this test you're basing the test on the config having the default value, or the default of the function having that value. If the config or the function change then this test now fails when it shouldn't. Since the -100 value is arbitrary.
2. It looks like you're just checking if you are able to have alignment between the attention mask and the label vector right?
3. You should test the inverse case also, that non padded positions have real labels. You can check like one or two of them. Because if MLM masking was setting everything to -100 this test would pass even though that would be incorrect behavior.

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+    
+    train_iterator, epoch_len = create_genomic_dataloader(args, tokenizer)
+    batch = next(train_iterator)
+    
+    assert batch["input_ids"].shape[0] == 3
+    assert batch["labels"].shape[0] == 3
+    assert batch["attention_mask"].shape[0] == 3
+
+
+def test_labels_equal_input_ids_for_causal_lm(simple_database, tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_dataloader.py]: You might come up with a better name for this test. It's just checking that when you mask out labels and remove the -100 value, you are placing the value of the corresponding input_id into that slot right?

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_tokenizer.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+@pytest.fixture
+def tokenizer():
+    """Load the ASCII nucleotide tokenizer."""
+    tokenizer_path = Path(__file__).parent.parent / "nucleotide_tokenizer"
+    return AutoTokenizer.from_pretrained(str(tokenizer_path))
+
+
+def test_tokenizer_special_token_ids(tokenizer):
+    """Test that the tokenizer's special token IDs are correct (match NeMo)"""
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_tokenizer.py:22]: these are different from what bionemo recipes is using right?

Comment (savitha-eng) [reply bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_tokenizer.py:22]: yes, i think the other recipes use tokenizers specific to the model (i.e., esm2 uses a amino acid tokenizer w a vocab size of 33, not ASCII tokenizers) so from a quick search, i didn't see anything that could be easily repurposed

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tokenizer_hf_fast.py [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+    
+    # Create tokenizer with NeMo convention (but we add BOS for causal LM)
+    tokenizer = create_nucleotide_tokenizer(
+        eos_id=0,
+        pad_id=1,
+        bos_id=2,  # Add BOS for causal language modeling
+        unk_id=3,
+    )
+    
+    logger.info("Tokenizer created")
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tokenizer_hf_fast.py:130]: you should get rid of the logs inside the tests.

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    # First sequence: 4 real + 4 padding
+    expected_mask_0 = [1, 1, 1, 1, 0, 0, 0, 0]
+    assert batch['attention_mask'][0].tolist() == expected_mask_0
+    
+    # Second sequence: all real
+    expected_mask_1 = [1, 1, 1, 1, 1, 1, 1, 1]
+    assert batch['attention_mask'][1].tolist() == expected_mask_1
+
+
+def test_tokenizer_vocab_size(tokenizer):
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_tokenizer.py]: kind of a funny test

File: bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_windowing.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        seq_length=1000,
+        tokenizer=tokenizer,
+        stride=1000,  # No overlap
+        min_window_length=500,
+        seed=42,
+    )
+    
+    assert len(dataset_overlap) == 29  # With overlap
+    assert len(dataset_no_overlap) == 25  # Without overlap
+    assert len(dataset_overlap) > len(dataset_no_overlap)
```
Comment (jomitchellnv) [inline bionemo-recipes/recipes/llama_native_te_nvfsdp/tests/test_windowing.py]: can you check the diff? len(d1) - len(d2) == 4
